src/core/mac/sub_mac.cpp

```diff
@@ -730,5 +794,130 @@ const char *SubMac::StateToString(State aState)
 
 // LCOV_EXCL_STOP
 
+//---------------------------------------------------------------------------------------------------------------------
+// CSL Receiver methods
+
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+void SubMac::SetCslChannel(uint8_t aChannel)
+{
+    mCslChannel = aChannel;
+}
+
+void SubMac::SetCslPeriod(uint16_t aPeriod)
+{
+    mCslPeriod = aPeriod;
+}
+
+void SubMac::SetCslTimeout(uint32_t aTimeout)
+{
+    mCslTimeout = aTimeout;
+}
+
+void SubMac::FillCsl(Frame &aFrame)
+{
+    uint8_t *cur = aFrame.GetHeaderIe(Frame::kHeaderIeCsl);
+
+    if (cur != NULL)
+    {
+        CslIe *csl = reinterpret_cast<CslIe *>(cur + sizeof(HeaderIe));
+
+        csl->SetPeriod(mCslPeriod);
+        csl->SetPhase(GetCslPhase());
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+        otLogDebgMac("%10lu:%s seq=%u phase=%hu", static_cast<uint32_t>(otPlatTimeGet()), __func__,
+                     aFrame.GetSequence(), csl->GetPhase());
+#endif
+    }
+}
+
+void SubMac::HandleCslTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<SubMac>().HandleCslTimer();
+}
+
+void SubMac::HandleCslTimer(void)
+{
+    switch (mCslState)
+    {
+    case kCslSample:
+    {
+        mCslState = kCslSleep;
+        IgnoreError(Sleep());
+        // OT_US_PER_TEN_SYMBOLS: computing CSL Phase using floor division.
+        mCslTimer.StartAt(mCslSampleTime, mCslPeriod * OT_US_PER_TEN_SYMBOLS - OT_US_PER_TEN_SYMBOLS);
+        otLogDebgMac("CSL sleep %u", mCslTimer.GetNow().GetValue());
+        break;
+    }
+
+    case kCslSleep:
+    {
+        mCslState = kCslSample;
+
+        mCslSampleTime += mCslPeriod * OT_US_PER_TEN_SYMBOLS;
+        otPlatRadioUpdateCslSampleTime(mCslSampleTime.GetValue());
+
+        // No need to call this when receiving or transmitting.
+        if (mState == kStateSleep)
+        {
+            IgnoreError(otPlatRadioReceive(&GetInstance(), mCslChannel));
+        }
+
+        mCslTimer.StartAt(mCslSampleTime, kCslSampleWindow * OT_US_PER_TEN_SYMBOLS);
+        otLogDebgMac("CSL Sample %u", mCslTimer.GetNow().GetValue());
+        break;
+    }
+
+    case kCslIdle:
+        break;
+
+    default:
+        OT_ASSERT(false);
+        break;
+    }
+}
+
+otError SubMac::StartCsl(void)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(mCslPeriod > 0, error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(!IsCslStarted(), error = OT_ERROR_ALREADY);
+
+    mCslSampleTime = mCslTimer.GetNow();
+    otPlatRadioUpdateCslSampleTime(mCslSampleTime.GetValue());
+
+    IgnoreError(Receive(mCslChannel));
+    mCslState = kCslSample;
+    mCslTimer.StartAt(mCslSampleTime, kCslSampleWindow * OT_US_PER_TEN_SYMBOLS);
+
+exit:
+    return error;
+}
+
+void SubMac::StopCsl(void)
+{
+    VerifyOrExit(mCslState != kCslIdle, OT_NOOP);
+
+    mCslTimer.Stop();
+    mCslState = kCslIdle;
+
+exit:
+    return;
+}
+
+uint16_t SubMac::GetCslPhase(void) const
+{
+    TimeMicro now  = TimerMicro::GetNow();
```

Comment:
use `otPlatRadioGetNow`? The CSL phase should be using radio's timer reference.

Reply:
I think they should always be the same in platform implementation. For example, on nrf528xx, `otPlatRadioGetNow` returns microsecond time and it just calls `otPlatTimeGet` which uses `nrf5AlarmGetCurrentTime` to get time. And `TimerMicro::GetNow` calls `otPlatAlarmMicroGetNow` which also uses `nrf5AlarmGetCurrentTime` to get time.

Here the `TimerMicro::GetNow` is used because the time would be passed to the timer. Using this method to get time seems more convinient to me. And since it's in `SubMac`, I think the underlying time is always the same thing for `SubMac` and radio, even for RCP mode.